File: tools/pageless-pdf/stack.py
#!/usr/bin/env python3
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Stack the chunk pages pageless_pdf.lua shipped into one strip page.

    stack.py <job.pdf> <pageless.json> <out.pdf>

Each chunk page becomes a Form XObject placed edge to edge down a single page
whose height is the sum of the chunk pages' MediaBox heights — the heights
LuaTeX wrote, so the stacking is exact to the backend's own rounding. Captured
TikZ pictures, which the galley holds as empty placeholder boxes, are drawn
back from their private pages at the recorded positions. The result is one
page, as tall as the document, that a PDF rasteriser can crop from at any
offset (pdftoppm -x -y -W -H): the reference rendering of the pageless
document, for pixel comparison against the browser.
"""
import json
import sys
from pathlib import Path

import pikepdf
from pikepdf import Name

SP_PER_BP = 65536 * 72.27 / 72          # TeX sp per PostScript point


def main() -> None:
    if len(sys.argv) != 4:
        sys.exit(__doc__)
    job, meta_path, out_path = Path(sys.argv[1]), Path(sys.argv[2]), Path(sys.argv[3])
    meta = json.loads(meta_path.read_text())
    if meta.get('unit') != 'sp':
        sys.exit('unexpected unit in pageless.json')

    src = pikepdf.open(job)
    chunks = meta['chunks']
    if not chunks:
        sys.exit('no chunks recorded')

    # MediaBox heights as LuaTeX wrote them, so the strip is consistent with
    # the pages' own content placement
    heights = []
    width = None
    for ch in chunks:
        page = src.pages[ch['page'] - 1]
        x0, y0, x1, y1 = (float(v) for v in page.mediabox)
        heights.append(y1 - y0)
        width = x1 - x0 if width is None else width
    total = sum(heights)

    out = pikepdf.new()
    # Not add_blank_page: it enforces Acrobat's 14400-unit page limit, which
    # a document strip exceeds by design. Poppler and MuPDF render any size.
    strip = pikepdf.Page(out.make_indirect(pikepdf.Dictionary(
        Type=Name.Page,
        MediaBox=pikepdf.Array([0, 0, width, total]),
        Resources=pikepdf.Dictionary(),
    )))
    out.pages.append(strip)
    ops = []
    y_top = 0.0                             # from the top of the strip
    for ch, h in zip(chunks, heights):
        page = src.pages[ch['page'] - 1]
        xobj = out.copy_foreign(page.as_form_xobject())
        name = strip.add_resource(xobj, Name.XObject, prefix='Chunk')
        y = total - (y_top + h)             # PDF y of the chunk's bottom edge
        ops.append(f'q 1 0 0 1 0 {y:.6f} cm {name} Do Q')
        y_top += h

    drawn = 0
    for pic in meta.get('pictures', []):
        if not pic.get('page'):
            continue
        page = src.pages[pic['page'] - 1]
        xobj = out.copy_foreign(page.as_form_xobject())
        name = strip.add_resource(xobj, Name.XObject, prefix='Pic')
        x = pic['x'] / SP_PER_BP
        # the private page is exactly the box's size; its bottom edge sits at
        # the box's depth below the strip position of the box's top plus height
        y = total - (pic['y'] + pic['height'] + pic['depth']) / SP_PER_BP
        ops.append(f'q 1 0 0 1 {x:.6f} {y:.6f} cm {name} Do Q')
        drawn += 1

    strip.Contents = out.make_stream('\n'.join(ops).encode('ascii'))
    out.save(out_path)
    print(f'{out_path}: 1 page, {width:.3f} x {total:.3f} bp, '
          f'{len(chunks)} chunk(s), {drawn} picture(s)')


if __name__ == '__main__':
    main()

File: tools/pageless-pdf/stack.ts
#!/usr/bin/env node
// SPDX-License-Identifier: AGPL-3.0-or-later
// Stack the chunk pages pageless_pdf.lua shipped into one strip page.
//
//     node tools/pageless-pdf/stack.ts <job.pdf> <pageless.json> <out.pdf>
//
// Each chunk page becomes a Form XObject placed edge to edge down a single page
// as tall as the chunk pages together – the heights LuaTeX wrote, so the
// stacking is exact to the backend's own rounding. Captured TikZ pictures,
// which the galley holds as empty placeholder boxes, are drawn back from their
// private pages at the recorded positions. The result is one page, as tall as
// the document: the reference rendering of the pageless document, for
// comparison with the browser.
import { readFileSync, writeFileSync } from 'node:fs';
import { PDFDocument } from 'pdf-lib';

const SP_PER_BP = 65536 * 72.27 / 72;          // TeX sp per PostScript point

interface Meta {
  unit: string;
  chunks: { page: number }[];
  pictures?: { page?: number; x: number; y: number; height: number; depth: number }[];
}

/** Stack job.pdf's chunk pages; returns what was drawn. */
export async function stack(jobPdf: string, meta: Meta, outPdf: string): Promise<{ width: number; height: number; chunks: number; pictures: number }> {
  if (meta.unit !== 'sp') throw new Error('unexpected unit in pageless.json');
  if (!meta.chunks.length) throw new Error('no chunks recorded');
  const src = await PDFDocument.load(readFileSync(jobPdf));
  const out = await PDFDocument.create();
  // MediaBox heights as LuaTeX wrote them
  const heights = meta.chunks.map(c => src.getPage(c.page - 1).getMediaBox().height);
  const width = src.getPage(meta.chunks[0].page - 1).getMediaBox().width;
  const total = heights.reduce((a, b) => a + b, 0);
  // (pdf-lib has no page-size limit; Acrobat's 14400 units do not apply here)
  const strip = out.addPage([width, total]);
  const pics = (meta.pictures ?? []).filter(p => p.page);
  const embedded = await out.embedPdf(src, [...meta.chunks.map(c => c.page - 1), ...pics.map(p => p.page! - 1)]);
  let top = 0;
  meta.chunks.forEach((_, i) => {
    strip.drawPage(embedded[i], { x: 0, y: total - (top + heights[i]) });
    top += heights[i];
  });
  // a picture's private page is exactly its box: its bottom edge sits at the
  // box's depth below the strip position of the box's top plus height
  pics.forEach((p, k) => strip.drawPage(embedded[meta.chunks.length + k],
    { x: p.x / SP_PER_BP, y: total - (p.y + p.height + p.depth) / SP_PER_BP }));
  writeFileSync(outPdf, await out.save());
  return { width, height: total, chunks: meta.chunks.length, pictures: pics.length };
}

if (import.meta.main) {
  const [job, metaPath, outPath] = process.argv.slice(2);
  if (!outPath) { console.error('usage: stack.ts <job.pdf> <pageless.json> <out.pdf>'); process.exit(2); }
  const r = await stack(job, JSON.parse(readFileSync(metaPath, 'utf8')), outPath);
  console.log(`${outPath}: 1 page, ${r.width.toFixed(3)} x ${r.height.toFixed(3)} bp, ${r.chunks} chunk(s), ${r.pictures} picture(s)`);
}
